[Blueprint] Implement logging for export and import operations

chihsuan opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

<!-- For bug fixes: If known, please provide links to help with traceability and escape analysis. -->
<!-- Please include a link to the issue of the bug being fixed, if one doesn't exist please create it. -->
<!-- If the PR that introduced the bug is known, please also add its link below. -->

Closes WOOPLUG-3946

This PR adds logging for the Blueprint package to improve traceability and debugging capabilities during export and import operations. This aids in traceability and incident response.

The following things are logged:

1. User ID - Who is doing the export/import
2. Step names - Which steps are being exported/imported
3. Step classes - Which classes are being used to export/import the steps (which is useful to know the detail actions taken)
4. success or error - Whether the export/import was successful or not

### Key Changes

- Add new `Logger` class for centralized logging
- Add detailed logging for export operations
- Add detailed logging for import operations


<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. pnpm install
2. Enable the Blueprint feature `wp option set woocommerce_feature_blueprint_enabled yes`
3. Go to `/wp-admin/admin.php?page=wc-settings&tab=advanced&section=blueprint` and export a blueprint
4. Go to WooCommerce > Status > Logs and check the logs for the Blueprint logs
5. Confirm that it logs the export steps
6. Import the blueprint
7. Confirm that it logs the import steps

![Screenshot 2025-04-18 at 14 48 41](https://github.com/user-attachments/assets/eb61a76c-47a8-4d03-a71e-defa570f7c47)
![Screenshot 2025-04-18 at 14 47 42](https://github.com/user-attachments/assets/70fe1cf3-b754-482f-bc52-f02ae0d689ac)


<!-- End testing instructions -->

### Testing that has already taken place:

<!-- Detail any testing that has already been conducted. -->
<!-- Include environment details such as hosting type, plugins, theme, store size, store age, and relevant settings. -->
<!-- Mention any analysis performed, such as assessing potential impacts on environment attributes and other plugins, performance profiling, or LLM/AI-based analysis. -->
<!-- Within the testing details you provide, please ensure that no sensitive information (such as API keys, passwords, user data, etc.) is included in this public pull request. -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

chihsuan committed:
Implement logging functionality in ExportSchema and ImportStep classes

- Introduced a Logger class to handle logging of export and import steps, improving traceability and debugging.
- Enhanced ExportSchema to log the start and completion of export steps, as well as errors encountered during the process.
- Updated ImportStep to log the start and result of the import process, including success and error messages.
- Refactored error handling in both classes to utilize the new logging system, ensuring consistent logging practices across the codebase.

chihsuan committed:
Add changelog

chihsuan committed:
Fix tests

chihsuan committed:
Refactor Logger class to utilize WP functions

- Introduced UseWPFunctions trait to encapsulate WordPress function usage.
- Updated Logger class to use wp_get_current_user_id() for retrieving the current user's ID, enhancing consistency with WordPress standards.
- Minor adjustment in ImportStepTest class to ensure proper namespace usage for WP_UnitTestCase.

louwie17 commented on packages/php/blueprint/src/ImportStep.php:
```diff
@@ -80,33 +82,80 @@ function ( $key, $importer ) {
 		if ( ! $importer instanceof StepProcessor ) {
-			$result->add_warn( "Importer {$this->step_definition->step} is not a valid step processor" );
-			return $result;
+			$result->add_error( 'Incorrect importer type' );
```
Any specific reason for changing this and the one above to `error` instead of `warn`?

chihsuan replied:
I realized that using `warn` would be problematic for this case because it wouldn't be visible in the UI, and the import would be considered successful, but it wouldn't import anything.

louwie17 commented on packages/php/blueprint/src/ImportStep.php:
```diff
@@ -80,33 +82,80 @@ function ( $key, $importer ) {
+	 *
+	 * @return bool True if the step can be imported, false otherwise.
+	 */
+	protected function can_import( &$result ) {
```
Should we log these errors as well, if the user can't import for example?

chihsuan replied:
I think additional error logging is unnecessary, as users receive error messages in UI or CLI with `--show-messages` when imports fail, which should be a better experience. Logging is the primary focus of security-related events. Let me know if you don't see any errors when permissions are not allowed. 🙏

chihsuan replied:
Ah, there is a known issue that errors are not shown correctly. It's fixed in https://github.com/woocommerce/woocommerce/pull/57336.

louwie17 commented on packages/php/blueprint/src/ExportSchema.php:
```diff
@@ -93,26 +96,57 @@ function ( $exporter ) {
+
+			$step_name           = $exporter instanceof HasAlias ? $exporter->get_alias() : $exporter->get_step_name();
+			$export_steps[]      = $step_name;
+			$exporters_classes[] = get_class( $exporter );
```
I think we should move this stuff to the `Logger` class.
For example, we could have a `$logger->start_export` function that takes the list of `$exporters`. This would avoid us having to explicitly getting the name and class name here, and instead move that to the Logger class.
We could have similar functions for `export_step_success` and `start_import` etc...

chihsuan replied:
Sounds good! I'll move this to the Logger class.

chihsuan replied:
Updated ExportSchema and ImportStep to use new logging methods in the Logger class in 4d0c0784e68ecc9454860abbf07ceac546ddab19

louwie17 reviewed (commented):
This tested quite well, thanks! I did notice if the user doesn't have permissions it doesn't log anything.
I also left several inline comments, let me know what you think about them?
Lastly, I kinda wonder if we should do the same thing for the export as we did for the import, where we do log when each step is done, instead of a single log line for all of the Export steps.

chihsuan commented:
> Lastly, I kinda wonder if we should do the same thing for the export as we did for the import, where we do log when each step is done, instead of a single log line for all of the Export steps.

I had the same thought about logging individual export steps. However, I found that logging each export step created excessive verbosity in the logs, making them harder to read. Unlike imports, which import data at each step, exports only complete when all steps succeed. Therefore, a single log entry at the end should be sufficient. Let me know if you think otherwise. 🙏

chihsuan committed:
Refactor logging in ExportSchema and ImportStep classes

- Updated ExportSchema and ImportStep to utilize new logging methods in the Logger class for starting, completing, and handling failures of export and import steps.
- Removed redundant logging code and centralized logging functionality in the Logger class, enhancing code clarity and maintainability.

chihsuan committed:
Remove WC_Log_Levels and WC_Logger_Interface stubs; centralize logging in tests

- Deleted the WC_Log_Levels and WC_Logger_Interface stub files to streamline the codebase.
- Introduced a new stubs.php file to provide necessary stubs for testing, ensuring compatibility and reducing redundancy.
- Updated references in the Logger class to use fully qualified class names for WC_Log_Levels, enhancing clarity and consistency.

louwie17 commented on packages/php/blueprint/src/ExportSchema.php:
```diff
@@ -100,19 +102,24 @@ function ( $exporter ) {
-				$schema['steps'][] = $step->get_json_array();
+			} catch ( \Throwable $e ) {
+				$step_name = $exporter instanceof HasAlias ? $exporter->get_alias() : $exporter->get_step_name();
+				$logger->export_step_failed( $step_name, $e );
```
May be nice to just pass the `$exporter` to the `step_failed` and also move the `$step_name` logic to the Logger.
This is small though, so not a blocker to merge.

louwie17 reviewed (approved):
Nice work on this and thanks for the updates!
This tested well and code looks good! I left one tiny comment, but not a blocker to merge.
